bluetooth/src/commonMain/kotlin/device/DefaultDeviceConnectionManager.kt

```diff
@@ -71,54 +57,98 @@ abstract class BaseDeviceConnectionManager(
+            coroutineScope: CoroutineScope
+        ): BaseDeviceConnectionManager
+    }
+
+    private val logTag = "Bluetooth Device ${deviceWrapper.identifier.stringValue}"
+
     private val _currentAction = AtomicReference<DeviceAction?>(null)
     protected var currentAction: DeviceAction?
         get() = _currentAction.get()
         set(value) { _currentAction.set(value) }
     protected val notifyingCharacteristics = sharedMutableMapOf<String, Characteristic>()
 
-    private val sharedEvents = SequentialMutableSharedFlow<Event>(0, bufferCapacity, coroutineScope)
-    val events = sharedEvents.asSharedFlow()
+    private val sharedEvents = SequentialMutableSharedFlow<DeviceConnectionManager.Event>(0, settings.eventBufferSize, coroutineScope)
+    override val events = sharedEvents.asSharedFlow()
 
     private val sharedRssi = MutableSharedFlow<Int>(0, 1, BufferOverflow.DROP_OLDEST)
-    val rssi = sharedRssi.asSharedFlow()
-
-    abstract fun getCurrentState(): State
+    override val rssi = sharedRssi.asSharedFlow()
 
-    abstract suspend fun connect()
-    abstract suspend fun discoverServices()
-    abstract suspend fun disconnect()
-    abstract suspend fun readRssi()
-    abstract suspend fun requestMtu(mtu: Int): Boolean
-    abstract suspend fun performAction(action: DeviceAction)
+    override suspend fun readRssi() {
+        logDebug { "Request Read RSSI" }
+    }
 
     fun handleNewRssi(rssi: Int) {
+        logDebug { "Updated Rssi $rssi" }
         sharedRssi.tryEmit(rssi)
     }
 
     fun handleNewMtu(mtu: Int) {
-        sharedEvents.tryEmitOrLaunchAndEmit(Event.MtuUpdated(mtu))
+        logDebug { "Updated Mtu $mtu" }
+        emitSharedEvent(DeviceConnectionManager.Event.MtuUpdated(mtu))
     }
 
-    fun startConnecting() {
-        sharedEvents.tryEmitOrLaunchAndEmit(Event.Connecting)
+    override fun startConnecting() {
+        logInfo { "Start Connecting" }
+        emitSharedEvent(DeviceConnectionManager.Event.Connecting)
     }
 
-    fun cancelConnecting() {
-        sharedEvents.tryEmitOrLaunchAndEmit(Event.CancelledConnecting)
+    override fun cancelConnecting() {
+        logInfo { "Cancel Connecting" }
+        emitSharedEvent(DeviceConnectionManager.Event.CancelledConnecting)
     }
 
-    fun handleConnect() {
-        sharedEvents.tryEmitOrLaunchAndEmit(Event.Connected)
+    override fun handleConnect() {
+        logInfo { "Did Connect" }
+        emitSharedEvent(DeviceConnectionManager.Event.Connected)
     }
 
-    fun startDisconnecting() {
-        sharedEvents.tryEmitOrLaunchAndEmit(Event.Disconnecting)
+    override fun startDisconnecting() {
+        logInfo { "Start Disconnecting" }
+        emitSharedEvent(DeviceConnectionManager.Event.Disconnecting)
     }
 
-    fun createService(wrapper: ServiceWrapper): Service = Service(wrapper, sharedEvents)
+    override suspend fun performAction(action: DeviceAction) {
+        logInfo { "Perform action $action" }
```

Comment:
This is strange, this abstract class does some work in some override methods for `DeviceConnectionManager` interface and some are just empty. So compiler will not warn about missing method overload and it will do nothing (only log). What is the rule to override methods and do work (all the time)?

Reply:
it should call another abstract method or something, I added a TODO for this.